Lib/ntpath.py

```diff
@@ -146,29 +146,50 @@ def splitdrive(p):
             sep = b'\\'
             altsep = b'/'
             colon = b':'
+            prefix = b'\\\\?\\'
+            unc_prefix = b'UNC\\'
         else:
             sep = '\\'
             altsep = '/'
             colon = ':'
+            prefix = '\\\\?\\'
+            unc_prefix = 'UNC\\'
         normp = p.replace(altsep, sep)
-        if (normp[0:2] == sep*2) and (normp[2:3] != sep):
-            # is a UNC path:
-            # vvvvvvvvvvvvvvvvvvvv drive letter or UNC path
-            # \\machine\mountpoint\directory\etc\...
-            #           directory ^^^^^^^^^^^^^^^
-            index = normp.find(sep, 2)
-            if index == -1:
+        if normp[:4] == prefix:
+            if normp[4:8] == unc_prefix:
+                # e.g. \\?\UNC\server\share\dir\file
+                start = 8
+            elif normp[5:6] == colon and not normp[6:7].strip(sep):
+                # e.g. \\?\c:\dir\file
+                return p[:6], p[6:]
+            else:
+                # anything else with a \\?\ prefix
```

Comment:
This excludes volume GUID paths, which are supported by the Windows path API and currently supported by `splitdrive()` as UNC paths.

GUID volume names are used for volume mountpoints. For example:

```
>>> os.readlink('C:\\Mount\\SystemPartition')
'\\\\?\\Volume{f2b8bcce-0000-0000-0000-a0e013000000}\\'
```
and by the volume management API:
```
>>> win32file.GetVolumePathNamesForVolumeName('\\\\?\\Volume{f2b8bcce-0000-0000-0000-a0e013000000}\\')
['C:\\Mount\\SystemPartition\\']
```
They can be returned by `GetFinalPathNameByHandleW()`. For example:

```
>>> f = open('C:\\Mount\\SystemPartition\\spam.txt', 'w')
>>> h = msvcrt.get_osfhandle(f.fileno())
>>> ntpath.splitdrive(win32file.GetFinalPathNameByHandle(h, 1))
('\\\\?\\Volume{f2b8bcce-0000-0000-0000-a0e013000000}', '\\spam.txt')
```

When checking for a GUID volume name, WinAPI `PathCchSkipRoot()` validates the hex digits and uses a case-insensitive comparison for "Volume". For example, `nt._path_splitroot()` calls `PathCchSkipRoot()`:

valid:
```
>>> nt._path_splitroot(r'\\?\VOLUME{00000000-0000-0000-0000-000000000000}\spam')
('\\\\?\\VOLUME{00000000-0000-0000-0000-000000000000}\\', 'spam')
```
invalid:
```
>>> nt._path_splitroot(r'\\?\VOLUME{00000000-0000-0000-0000-00000000000G}\spam')
('', '\\\\?\\VOLUME{00000000-0000-0000-0000-00000000000G}\\spam')
```

Reply:
I've addressed this, except for the validation of the hex guid. It was buggy before, and it would unfair to future Python contributors if we fixed _all_ the bugs here ;)

Reply:
It's fine; just a minor suggestion to improve the reliability. 

As to the old result, the generic match of any device name as a 'drive' is still possible with paths in the "." device namespace, such as "\\\\.\\PIPE\spampipe". You're only restricting the "?" filesystem namespace.

Reply:
To me it feels like progress - we've fixed some cases, and removed a duplicate implementation in pathlib, without causing regressions. I see your point that there's more to do, though

Reply:
I didn't mean to imply it's not progress. I've just been reluctant to adopt the Windows path API's imposed limit that the device name for a valid "\\\\?\\" must be "UNC", a drive letter, or a volume GUID name. 

The "\\\\?\\" prefix disables normalization when the path is opened, whereas "\\\\.\\" device paths are always normalized. Ultimately, both prefixes get mapped to NT's "\\??\\" device prefix [\*] before making an NT system call such as `NtCreateFile()` or `NtOpenFile()`.  Bypassing normalization allows passing literal forward slashes and repeated slashes in the rest of the path. It also allows the path to end with spaces or dots. Also, if long paths aren't enabled for the process, skipping normalization allows opening the native maximum path length, up to about 32,760 characters. Since every device has its own namespace, there could be a valid reason to use a "\\\\?\\" literal/extended path with some device other than a DOS drive or GUID drive. 

I'm concerned that the lack of a perceived 'drive' and/or root path can lead to such paths getting processed as relative paths and resolved against the current working directory in some cases. A UNC path such as "\\\\?\\SpamDevice\\Eggs" is always an absolute path. I understand that if such paths aren't supported, then undefined behavior like this is possible and allowed, so I was reluctant to change `splitdrive()` to return `('', '\\\\?\\SpamDevice\\Eggs')` instead of `('\\\\?\\SpamDevice', '\\Eggs')`.

---
[\*] NT's "\\??\\" prefix is shorthand for the device directory of the current thread, based on the logon session of the thread's security context. If the thread doesn't have an impersonation token for another security context, then it's based on the logon session of the process token. Each logon-session device directory is linked to and shadows the global device directory, "\\GLOBAL??", except the SYSTEM logon session directly uses the global device directory. These so-called 'device' directories actually contain NT object symlinks to device objects in "\\Device". Using object symlinks provides flexibility and allows some DOS 'devices' to actually be paths on a device, such as mapped and substitute drives (e.g. "\\??\\W:" -> "\\??\\C:\Windows"), or "\\??\\CONIN$" and "\\??\\CONOUT$", which resolve to "\\Device\\ConDrv\\CurrentIn" and "\\Device\\ConDrv\\CurrentOut" on the "ConDrv" device.